templates/README.md

```diff
@@ -0,0 +1,254 @@
+# Template development guidelines
+
+The idea is to reuse and enhance the concept of OpenShift templates and its parameter expansion and also to reuse the kubevirt’s VM.spec structure as a DTO that will provide data to the parameter expansion.
+
+Since the templates will contain lots of redundancies, there should be a build time templating mechanism for composing the final templates from snippets.
+
+Also, there might be multiple different OSes, flavors, sizes mentioned by any single template file if they share the same configuration.
+
+Please note that the cnv.io suffix used in labels and annotations is temporary and is likely to change.
+
+## User Experience
+
+The flow should be the same as usual with regards to openshift templates. The user either takes a local template and generates a VM object.
+
+Or the user gets a template from kubernetes and invokes the proper REST command on top of it to keep the VM linked to the template or edit it in accordance to the template.
+
+The UI can then operate on top of template parameters and/or the Template.VM.spec DOM directly. There is an openshift [REST endpoint for instantiating Templates](https://docs.openshift.com/container-platform/3.7/rest_api/examples.html#template-instantiation) and the UI might be able to use it to create an in memory instance of VM it will then work with before posting it (kubernetes dry-run issues [https://github.com/kubernetes/kubernetes/issues/11488](#11488) and [https://github.com/kubernetes/kubernetes/issues/63559](#63559)).
+
+## Example file
+
+### template.yaml
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: Template # Openshift kind of template or something "better"
+metadata:
+  name: windows-10
+  annotations:
+    openshift.io/display-name: "Generic Microsoft Windows 7+ template"
+    description: "Basic windows template long description with details"
+    openshift.io/long-description: >-
+      Long description of the template
+    openshift.io/provider-display-name: "Red Hat, Inc."
+    openshift.io/documentation-url: "https://cnv.io/..."
+    openshift.io/support-url: "https://access.redhat.com"
+    iconClass: icon-windows    
+
+    # This set of annotations is meant as a hint only, it is
+    # not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded here informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+    defaults.template.cnv.io/disk: default-disk
```

Comment:
what does this actually mean? The bootable disk?

Reply:
The disk entity you should use as a template for constructing all added disks. Basically prepare a disk in UI and add all specified fields from the template (like the bus: virtio for example). This is needed to support different default drivers on different OSes.

Reply:
Btw, we could also express this logic using the replacements that were proposed as extensions. It will just require specifying them now and might result in less readable template. And the disk there would have to be removed by the UI, but it is used in test automation.

Reply:
The is also default-volume, just two lines below :)

Reply:
And this is supposed to define what kind of disk is supported by the OS mainly in terms of bus (ide, scsi, sata, virtio, ...)

Reply:
would be best if it clearly states that it's the default guest device type. so..just your last comment transformed into the readme:)

Reply:
Good idea. I wonder if we need to distinguish the default template based on disk type - HDD vs. CD

Reply:
not really. That's not what you're commonly adding to VM. Sure, the driver may be different for Win vs Lin (SATA vs virtio-scsi) but you would put that straight to the template I suppose (as you'd want to create an empty CDROM all the time) and then the backend code which would add another one for e.g. cloud-init just need to use the same as the existing. I don't think that logic exists, but it's not too difficult.

Alternatively you could add a default for CDs...but...ideally we don't want to explode there either so unless it's really needed I would try to avoid introducing more

Reply:
okey, I've got your intention now :) So how would you express the list of available values in the default disk? e.g. it supports both virtio-scsi and ide, but the default is virtio-scsi?

Reply:
The disk specification can be quite different based on what disk, volume or nic is being used. The intention here is to just merge the specified template with what you create in the UI. If the OS supports both then the template will either prefer one or the ui is free to provide any value (or none to use kubevirt defaulting).

Reply:
To make it clear, the list of editable fields for disks is not going to be part of the template. The UI will need to have some knowledge about how disks are specified. The template is just supposed to be applied on top of what the UI prepares.

Reply:
so it's fine that the disk structure is going to be written by UI, that's expected, but for validation&display UI needs to know valid values for that OS/kubevirt version

Reply:
Right, we can prepare the validation rules for those. As soon as we figure out the format..